adapter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,312 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stackdriver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/googleapis/gax-go"
+	xcontext "golang.org/x/net/context"
+	gapiopts "google.golang.org/api/option"
+	labelpb "google.golang.org/genproto/googleapis/api/label"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"istio.io/mixer/adapter/stackdriver/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+// TODO: implement adapter validation
+// TODO: change batching to be size aware: right now we batch and send data to stackdriver based on only a ticker.
+// Ideally we'd also size our buffer and send data whenever we hit the size limit or config.push_interval time has passed
+// since the last push.
+// TODO: today we start a ticker per aspect instance, each keeping an independent data set it pushes to SD. This needs to
+// be promoted up to the factory, which will hold a buffer that all aspects write in to, with a single ticker/loop responsible
+// for pushing the data from all aspect instances.
+
+type (
+
+	// Abstracts the creation of the stackdriver client to enable network-less testing.
+	createClientFunc func(*config.Params) (*monitoring.MetricClient, error)
+
+	factory struct {
+		adapter.DefaultBuilder
+		createClient createClientFunc
+
+		once sync.Once
+		t    *time.Ticker // we hold on to a ref so we can stop it during Close()
+	}
+
+	sdinfo struct {
+		ttype string
+		kind  metricpb.MetricDescriptor_MetricKind
+		value metricpb.MetricDescriptor_ValueType
+	}
+
+	// Abstracts over client.CreateTimeSeries for testing
+	pushFunc func(ctx xcontext.Context, req *monitoringpb.CreateTimeSeriesRequest, opts ...gax.CallOption) error
+
+	sd struct {
+		// TODO: remove when env is request scoped
+		env adapter.Env // used for logging
+
+		projectID   string
+		metricInfo  map[string]sdinfo
+		client      *monitoring.MetricClient
+		pushMetrics pushFunc
+
+		// TODO: move this up to the factory so there's a single batch being sent to SD for the entire Mixer, rather than one per aspect instance.
+		// Stackdriver's SDK doesn't perform batching, though their API supports it.
+		// We'll roll our own batching by aggregating timeseries and periodically sending them to Stackdriver.
+		m      sync.Mutex // guards toSend
+		toSend []*monitoringpb.TimeSeries
+	}
+)
+
+const (
+	adapterName = "stackdriver"
+	adapterDesc = "Publishes StackDriver metricInfo, logs, and traces."
+
+	// From https://github.com/GoogleCloudPlatform/golang-samples/blob/master/monitoring/custommetric/custommetric.go
+	customMetricPrefix = "custom.googleapis.com/"
```

Comment:
This is not configurable, all custom metrics are required to use this prefix by the Stackdriver API.